### PR TITLE
OCPBUGS-10478: gather: azure: fix collecting VM serial logs

### PR DESCRIPTION
**Pull request by r4f4**
With the change to use managed storage accounts for boot diagnostics [1], we were getting empty filenames for the boot diagnostics files:

```
INFO Pulling VM console logs
DEBUG attemping to download
DEBUG attemping to download
DEBUG attemping to download
DEBUG attemping to download
DEBUG attemping to download
DEBUG attemping to download
DEBUG attemping to download
DEBUG attemping to download
INFO Failed to gather VM console logs: unable to download file: /c/serial-log-bundle-20230320085802
```

After adding some more debug information, we've found out the blob URIs
are just null:

```
DEBUG appending boot diags for vm rdossant-03200931-5rld5-bootstrap: &{<nil> <nil>}
DEBUG appending boot diags for vm rdossant-03200931-5rld5-master-0: &{<nil> <nil>}
DEBUG appending boot diags for vm rdossant-03200931-5rld5-master-1: &{<nil> <nil>}
DEBUG appending boot diags for vm rdossant-03200931-5rld5-master-2: &{<nil> <nil>}
```

That is because boot diagnostics data are not present in the InstanceView of the VirtualMachine but actually need to be retrieved separately when using managed storage. And support for doing that was added to compute profile version >= 2020-09-01. Once we have the diagnostics data URIs, they are generated with a SAS token and can be be downloaded directly.

So this PR introduces the following changes:

 * Update the code to use the new Azure SDK
 * If no diagnostics data is found in the InstanceView, then check the BootDiagnosticsData object
 * Check the file URI to figure out how to download it: if it starts with a owned storage account host, then use shared keys; otherwise, just download it with a simple request
  * Skip shared keys not matching the blob's storage account

[1] https://github.com/openshift/installer/pull/6848

### CODE REVIEW

**Review by MayXuQQ — APPROVED** [review]
test private cluster and public cluster, gather VM log succeed.

**Comment by MayXuQQ** [comment]
test with pre-merge build registry.build05.ci.openshift.org/ci-ln-7f6zsrt/release:latest 
 [build info](https://storage.googleapis.com/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1637871604962693120/build-log.txt) 
[pri_openshift_install.log](https://github.com/openshift/installer/files/11025282/pri_openshift_install.log)
[pub_openshift_install.log](https://github.com/openshift/installer/files/11025283/pub_openshift_install.log)

**Comment by r4f4** [comment]
/hold
This needs to be fixed to work with both Managed and Owned accounts.

**Comment by r4f4** [comment]
Updated the changes to: 
1. Transition to the new SDK in a separate commit
2. Fix getting boot logs from managed accounts (OCPBUGS-10478)
3. Handle downloading boot logs from either owned accounts (AzureStack) and Managed Accounts (Azure Public and Gov)

**NOTE**: I tried testing this in an AzureStack WWT environment but there seems to be a bug getting the blobs. It fails even without my changes, so it's probably a bug in `azure-sdk-for-go`. Reported it at https://github.com/Azure/azure-sdk-for-go/issues/20456

**Comment by r4f4** [comment]
Small update to fix lint issues.

**Review by patrickdillon — COMMENTED** [review]
/approve

This LGTM but looks like we are still waiting on tests.

>NOTE: I tried testing this in an AzureStack WWT environment but there seems to be a bug getting the blobs. It fails even without my changes, so it's probably a bug in azure-sdk-for-go. Reported it at https://github.com/Azure/azure-sdk-for-go/issues/20456

It seems like this is a known issue on Azure Stack. I have added a note in the release note text of OCPBUGS-10478. @bscott-rh do we need to do anything else for this to be listed as a known issue?

File: pkg/gather/azure/azure.go [hunk 1]
```diff
@@ -174,28 +179,51 @@ func getVirtualMachines(ctx context.Context, g *Gather) ([]*armcompute.VirtualMa
...
+			sshotURI = instanceView.BootDiagnostics.ConsoleScreenshotBlobURI
+			slogURI = instanceView.BootDiagnostics.SerialConsoleLogBlobURI
+		}
+
+		// Boot logs might be in managed account
+		if sshotURI == nil && slogURI == nil {
+			bootData, err := g.virtualMachinesClient.RetrieveBootDiagnosticsData(ctx, g.resourceGroupName, *vm.Name, nil)
 			if err != nil {
-				g.logger.Debugf("failed to get instance view: %v", err)
+				g.logger.Debugf("failed to get boot diagnostics data: %v", err)
```
[the remark this answers is not in the document]
Comment (r4f4) [reply pkg/gather/azure/azure.go:208]: Is it ok if I fix all of them in a commit?

**Comment by bscott-rh** [comment]
@patrickdillon spoke with a member of the release notes team. it should get picked up if you set "Release Note Type" to "Known Issue" in the OCPBUGS ticket.

**Comment by MayXuQQ** [comment]
pre-merge test failed 
[version registry.build05.ci.openshift.org/ci-ln-c52ct6k/release:latest ](https://storage.googleapis.com/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1638717984727896064/build-log.txt)

> INFO Credentials loaded from file "/var/home/core/.azure/osServicePrincipal.json" 
INFO Pulling VM console logs                      
INFO Failed to gather VM console logs: could not find any storage accounts: GET https://management.azure.com/subscriptions/53b8f551-f0fc-4bea-8cba-6d1fefd54c8a/resourceGroups/maxu-az-pri-ssfbg-rg/providers/Microsoft.Storage/storageAccounts 
INFO -------------------------------------------------------------------------------- 
INFO RESPONSE 404: 404 Not Found                  
INFO ERROR CODE: InvalidResourceType              
INFO -------------------------------------------------------------------------------- 
INFO {                                            
INFO   "error": {                                 
INFO     "code": "InvalidResourceType",           
INFO     "message": "The resource type 'storageAccounts' could not be found in the namespace 'Microsoft.Storage' for api version '2019-03-01'. The supported api-versions are '2023-01-01,2022-09-01,2022-05-01,2021-09-01,2021-08-01,2021-06-01,2021-05-01,2021-04-01,2021-02-01,2021-01-01,2020-08-01-preview,2019-06-01,2019-04-01,2018-11-01,2018-07-01,2018-03-01-preview,2018-02-01,2017-10-01,2017-06-01,2016-12-01,2016-05-01,2016-01-01,2015-06-15,2015-05-01-preview'." 
INFO   }                                          
INFO }          

try https://management.azure.com/subscriptions/53b8f551-f0fc-4bea-8cba-6d1fefd54c8a/resourceGroups/maxu-az-pri-ssfbg-rg/providers/Microsoft.Storage/storageAccounts

> {"error":{"code":"MissingApiVersionParameter","message":"The api-version query parameter (?api-version=) is required for all requests."}}

**Comment by r4f4** [comment]
> 

Interesting. When using an account client without specifying the APIVersion on WWT I get:
```
 POST https://management.mtcazs.wwtatc.com/subscriptions/d751283a-64fa-401b-92a1-58f1750ac0a7/resourceGroups/test-serial-log/providers/Microsoft.Storage/storageAccounts/seriallogaccnt/listKeys
--------------------------------------------------------------------------------
RESPONSE 400: 400 Bad Request
ERROR CODE: NoRegisteredProviderFound
--------------------------------------------------------------------------------
{
  "error": {
    "code": "NoRegisteredProviderFound",
    "message": "No registered resource provider found for location 'mtcazs' and API version '2022-09-01' for type 'storageaccounts'. The supported api-versions are '2019-06-01, 2019-04-01, 2018-11-01, 2018-07-01, 2018-02-01, 2017-10-01, 2017-06-01, 2016-12-01, 2016-05-01, 2016-01-01, 2015-06-15, 2015-05-01-preview'. The supported locations are 'mtcazs'."
  }
}
--------------------------------------------------------------------------------
```
So hopefully `2019-04-01` will work for all.

**Review by MayXuQQ — CHANGES_REQUESTED** [review]
pre-merge test pass

**Pull request merged**